### PR TITLE
Prometheus Metrics

### PR DESCRIPTION
**Pull request by arjunrn**
Following changes were made:

1. Removed existing metrics which was deadcode. It was not wired to anything and metrics were never collected. Also calling the API just to check if it's that exhausts the rate limits.
2. Added metric for operator reconcile duration time.
3. Added metric for API calls to the Dead Man's Snitch API
4. Added metric for API calls to the API server. Mostly copied from this [PR](https://github.com/openshift/aws-account-operator/pull/410) in [aws-account-operator](https://github.com/openshift/aws-account-operator/).

[Internal](https://issues.redhat.com/browse/OSD-4568) [Issues](https://issues.redhat.com/browse/OSD-4570)

### CODE REVIEW

File: cmd/manager/main.go [hunk 1]
```diff
@@ -117,14 +117,9 @@ func main() {
...
-	// Setup all Controllers
-	if err := controller.AddToManager(mgr); err != nil {
-		log.Error(err, "")
-		os.Exit(1)
-	}
-
+	localmetrics.Collector = localmetrics.NewMetricsCollector()
 	metricsServer := metrics.NewBuilder().WithPort(metricsPort).WithPath(metricsPath).
-		WithCollectors(localmetrics.MetricsList).
+		WithCollector(localmetrics.Collector).
```
Comment (fahlmant) [inline cmd/manager/main.go:122]: A Collector is a prometheus type that all different type of metrics implement. The object passed here has a collectors list within it, but `WithCollector` is just expecting one metric

Comment (arjunrn) [reply cmd/manager/main.go:122]: Right, so I implemented the `Collector` interface. Is there a problem with this approach?

Comment (fahlmant) [reply cmd/manager/main.go:122]: Would all future metrics, even non-reconcile or api related go in this collector?

Comment (2uasimojo) [reply cmd/manager/main.go:122]: AFAICT this is the pattern aws-account-operator uses.

https://github.com/openshift/aws-account-operator/blob/4bb28f759e24fdc03ccf943c9ec772150a421cd4/cmd/manager/main.go#L136
https://github.com/openshift/aws-account-operator/blob/4bb28f759e24fdc03ccf943c9ec772150a421cd4/pkg/localmetrics/localmetrics.go#L37-L55

seems fine by me.

File: pkg/controller/deadmanssnitch/deadmanssnitch_controller.go [hunk 2]
```diff
@@ -125,6 +126,11 @@ func (r *ReconcileDeadMansSnitch) Reconcile(request reconcile.Request) (reconcil
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	reqLogger.Info("Reconciling DeadMansSnitch")
 
+	start := time.Now()
+	defer func() {
+		localmetrics.Collector.ObserveReconcile(time.Since(start).Seconds())
```
Comment (2uasimojo) [inline pkg/controller/deadmanssnitch/deadmanssnitch_controller.go:131]: Other impls put a "Reconcile complete" log message in here that includes the duration. Probably not a must-have, but could be nice.

File: pkg/dmsclient/instrumented_client.go [hunk 3]
```diff
@@ -0,0 +1,125 @@
+package dmsclient
```
Comment (2uasimojo) [inline pkg/dmsclient/instrumented_client.go:1]: This file seems like a really heavy way to instrument the dms client. Why not simply wrap the guts of the `do` func? I can see it'll be a little bit tricky to get the `method` label in place for `FindSnitchesByName`, if we even care to distinguish that from `ListAll`, but it should be doable.

Comment (arjunrn) [reply pkg/dmsclient/instrumented_client.go:1]: IMO it's the same amount of complexity. In the `do()` method the path and payload will have to be inspected to determine which method is being called. I didn't want to change a lot of the existing code and instead chose to wrap it.

File: cmd/manager/main.go (hunk 1 above)
Comment (fahlmant) [reply cmd/manager/main.go:122]: huh, I didn't see that in the refactor. AFAIK, that's a pattern for proxying metrics from other monitoring stores and exporters, not for instrumenting metrics. But whatever, if it works

File: pkg/dmsclient/instrumented_client.go (hunk 3 above)
Comment (2uasimojo) [reply pkg/dmsclient/instrumented_client.go:1]: It's not about complexity -- I agree it would be doing the same thing either way, as it should. My concern is the code repetition, which carries with it the usual maintainability issues. There's also no need for the uninstrumented client to exist anymore.

File: pkg/dmsclient/dmsclient.go [hunk 4]
```diff
@@ -103,14 +108,20 @@ func (c *dmsClient) newRequest(method, path string, body interface{}) (*http.Req
 }
 
 func (c *dmsClient) do(req *http.Request) (*http.Response, error) {
+	start := time.Now()
```
Comment (2uasimojo) [inline pkg/dmsclient/dmsclient.go:111]: I like this implementation a lot better, thank you.

File: pkg/localmetrics/localmetrics.go [hunk 5]
```diff
@@ -16,51 +16,187 @@ package localmetrics
...
+// RecordSnitchCallDuration records the time taken to make a call to the Dead Man Snitch API
+func (m *MetricsCollector) RecordSnitchCallDuration(duration time.Duration, u *url.URL, method string) {
+	operation := m.parseSnitchCall(u.Path, method)
+	m.snitchCallDuration.With(prometheus.Labels{snitchMethodLabel: operation}).Observe(duration.Seconds())
+}
 
-	resp, err := http.DefaultClient.Do(req)
-	if err != nil {
-		MetricDeadMansSnitchHeartbeat.With(prometheus.Labels{"name": "deadmanssnitch-operator"}).Set(float64(0))
+func (m *MetricsCollector) parseSnitchCall(path, method string) string {
```
Comment (2uasimojo) [inline pkg/localmetrics/localmetrics.go:122]: It would be a bit more future-proof to add the `operation` as a `string` parameter to the `do` func. That way the label lives in the func it belongs to. (The only weird one would be `FindSnitchesByName`, which calls `ListAll` instead of `do`, but you're not handling that here anyway.)

File: pkg/dmsclient/dmsclient.go [hunk 6]
```diff
@@ -103,14 +108,20 @@ func (c *dmsClient) newRequest(method, path string, body interface{}) (*http.Req
...
+	defer func() {
+		c.metricsCollector.RecordSnitchCallDuration(time.Since(start), req.URL, req.Method)
```
[the remark this answers is not in the document]
Comment (arjunrn) [reply pkg/dmsclient/dmsclient.go:113]: I don't think the response status codes are very important. If there is an error counter and we see an increase we can investigate. Also I will add the note about the `_count`.

File: pkg/dmsclient/dmsclient.go (hunk 4 above)
Comment (arjunrn) [reply pkg/dmsclient/dmsclient.go:111]: :+1:

File: pkg/dmsclient/dmsclient.go [hunk 7]
```diff
@@ -147,7 +157,7 @@ func (c *dmsClient) List(snitchToken string) (Snitch, error) {
 		return snitch, err
 	}
 
-	resp, err := c.do(req)
+	resp, err := c.do(req, "describe")
```
Comment (2uasimojo) [inline pkg/dmsclient/dmsclient.go:160]: `"describe"` made me raise an eyebrow as it doesn't match the name of the func... but it's actually the name of the func that's questionable, since we're retrieving a single resource. +1

File: pkg/localmetrics/localmetrics.go [hunk 8]
```diff
@@ -16,51 +16,167 @@ package localmetrics
...
 )
 
-// UpdateMetrics updates all the metrics ever 5 minutes
-func UpdateMetrics() {
+type MetricsCollector struct {
+	ReconcileDuration  prometheus.Histogram
+	apiCallDuration    *prometheus.HistogramVec
+	snitchCallErrors   prometheus.Counter
+	snitchCallDuration *prometheus.HistogramVec
+	collectors         []prometheus.Collector
```
Comment (2uasimojo) [inline pkg/localmetrics/localmetrics.go:40]: Why is this here? I don't see it being used (except by `AddCollector`, which we also don't need, see below).

File: pkg/localmetrics/localmetrics.go [hunk 9]
```diff
@@ -16,51 +16,167 @@ package localmetrics
...
+
+// ObserveReconcile records the duration of the reconcile loop of the operator
+func (m *MetricsCollector) ObserveReconcile(seconds float64) {
+	m.ReconcileDuration.Observe(seconds)
+}
+
+// AddCollector adds a collector to existing metrics
+func (m *MetricsCollector) AddCollector(collector prometheus.Collector) {
+	m.collectors = append(m.collectors, collector)
+}
```
Comment (2uasimojo) [inline pkg/localmetrics/localmetrics.go:114]: Pretty sure we don't need this. It's unused, and `MetricsCollector` only needs to satisfy `prometheus.Collector` afaict, yes?

(If we need it in the future, we can always add it then. YAGNI.)

File: pkg/localmetrics/localmetrics.go [hunk 10]
```diff
@@ -16,51 +16,167 @@ package localmetrics
...
+
+// RecordSnitchCallDuration records the time taken to make a call to the Dead Man Snitch API
+func (m *MetricsCollector) RecordSnitchCallDuration(duration time.Duration, operation string) {
```
Comment (2uasimojo) [inline pkg/localmetrics/localmetrics.go:117]: nit: It would be nice to make these more consistent.
- Right now the names of the methods are `Add`, `Observe`, and `Record`. Can we pick one?
- We've got a `duration float64`, a `seconds float64`, and a `duration time.Duration` which does `.Seconds()` within the func. Can we pick one?

File: pkg/localmetrics/localmetrics.go [hunk 11]
```diff
@@ -16,51 +16,167 @@ package localmetrics
...
-	if err != nil {
-		MetricDeadMansSnitchHeartbeat.With(prometheus.Labels{"name": "deadmanssnitch-operator"}).Set(float64(0))
+func NewMetricsCollector() *MetricsCollector {
+	return &MetricsCollector{
+		ReconcileDuration: prometheus.NewHistogram(prometheus.HistogramOpts{
+			Name:        "dms_operator_reconcile_duration_seconds",
+			Help:        "The duration it takes to reconcile a ClusterDeployment",
+			ConstLabels: map[string]string{"name": operatorName},
+		}),
+		// apiCallDuration times API requests. Histogram also gives us a _count metric for free but also includes errors.
```
Comment (2uasimojo) [inline pkg/localmetrics/localmetrics.go:70]: Noticed after the fact (and I think this is my mistake, I think I flagged the wrong location in my previous comment): This isn't the right place for this comment, as apiCallDuration does *not* count errors. The comment needs to go next to `snitchCallDuration`.

I'm fixing this in the fup.

Comment (2uasimojo) [reply pkg/localmetrics/localmetrics.go:70]: Addressed in #69.

File: pkg/localmetrics/localmetrics.go (hunk 8 above)
Comment (2uasimojo) [reply pkg/localmetrics/localmetrics.go:40]: Addressed in #69.

File: pkg/localmetrics/localmetrics.go (hunk 9 above)
Comment (2uasimojo) [reply pkg/localmetrics/localmetrics.go:114]: Addressed in #69.

File: pkg/localmetrics/localmetrics.go (hunk 10 above)
Comment (2uasimojo) [reply pkg/localmetrics/localmetrics.go:117]: Addressed in #69.
